mqtt/mqtt-edgehub/src/auth/authorization/edgehub.rs

```diff
@@ -0,0 +1,336 @@
+
+        if is_iothub_topic(topic) {
+            // run authorization rules for publication to IoTHub topic
+            self.authorize_iothub_topic(client_id, client_info, topic)
+        } else if is_forbidden_topic(topic) {
+            Authorization::Forbidden(format!("{} is forbidden topic filter", topic))
+        } else {
+            // allow any client to publish to any non-iothub topics
+            Authorization::Allowed
+        }
+    }
+
+    fn authorize_subscribe(
+        &self,
+        client_id: &ClientId,
+        client_info: &ClientInfo,
+        subscribe: &Subscribe,
+    ) -> Authorization {
+        let topic = subscribe.topic_filter();
+
+        if is_iothub_topic(topic) {
+            // run authorization rules for subscription to IoTHub topic
+            self.authorize_iothub_topic(client_id, client_info, topic)
+        } else if is_forbidden_topic_filter(topic) {
+            Authorization::Forbidden(format!("{} is forbidden topic filter", topic))
+        } else {
+            // allow any client to subscribe to any non-iothub topics
+            Authorization::Allowed
+        }
+    }
+
+    fn authorize_iothub_topic(
+        &self,
+        client_id: &ClientId,
+        client_info: &ClientInfo,
+        topic: &str,
+    ) -> Authorization {
+        match client_info.auth_id() {
+            // forbid anonymous clients to subscribe to restricted topics
+            AuthId::Anonymous => Authorization::Forbidden(
+                "Anonymous clients do not have access to IoTHub topics".to_string(),
+            ),
+            // allow authenticated clients with client_id == auth_id and accessing its own IoTHub topic
+            AuthId::Identity(identity) if identity == client_id.as_str() => {
+                let allowed_topics = allowed_iothub_topic(client_id);
+                if allowed_topics
+                    .iter()
+                    .any(|allowed_topic| topic.starts_with(allowed_topic))
+                {
+                    Authorization::Allowed
+                } else {
+                    Authorization::Forbidden(
+                        "Client must connect to its own IoTHub topic".to_string(),
+                    )
+                }
+            }
+            // forbid access otherwise
+            AuthId::Identity(_) => Authorization::Forbidden(format!(
+                "client_id {} must match registered iothub identity id to access IoTHub topic",
+                client_id
+            )),
+        }
+    }
+}
+
+fn is_forbidden_topic_filter(topic_filter: &str) -> bool {
+    FORBIDDEN_TOPIC_FILTER_PREFIXES
+        .iter()
+        .any(|forbidden_topic| topic_filter.starts_with(forbidden_topic))
+}
+
+fn is_forbidden_topic(topic_filter: &str) -> bool {
+    topic_filter.starts_with('$')
+}
+
+fn is_iothub_topic(topic: &str) -> bool {
+    topic.starts_with("$edgehub/") || topic.starts_with("$iothub/")
+}
+
+fn allowed_iothub_topic(client_id: &ClientId) -> Vec<String> {
```

Comment:
Will this allocate for every pub? Can we do better? 
(regex or caching come to mind)